fe/fe-core/src/main/java/com/starrocks/metric/ResourceGroupMetricMgr.java

```diff
@@ -0,0 +1,232 @@
+
+            if (!RESOURCE_GROUP_QUERY_LATENCY_MAP.containsKey(resourceGroupName)) {
+                createQueryResourceGroupLatency(resourceGroupName);
+            }
+            List<GaugeMetricImpl> metricList = RESOURCE_GROUP_QUERY_LATENCY_MAP.get(resourceGroupName);
+            if (latencyList.size() > 0) {
+                metricList.get(0).setValue(latencySum / latencyList.size());
+
+                latencyList.sort(Comparator.naturalOrder());
+
+                int index = (int) Math.round((latencyList.size() - 1) * 0.5);
+                metricList.get(1).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.75);
+                metricList.get(2).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.90);
+                metricList.get(3).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.95);
+                metricList.get(4).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.99);
+                metricList.get(5).setValue((double) latencyList.get(index));
+                index = (int) Math.round((latencyList.size() - 1) * 0.999);
+                metricList.get(6).setValue((double) latencyList.get(index));
+            } else {
+                metricList.get(0).setValue(0.0);
+                metricList.get(1).setValue(0.0);
+                metricList.get(2).setValue(0.0);
+                metricList.get(3).setValue(0.0);
+                metricList.get(4).setValue(0.0);
+                metricList.get(5).setValue(0.0);
+                metricList.get(6).setValue(0.0);
+            }
+        }
+    }
+
+    private static void createQueryResourceGroupLatency(String resourceGroupName) {
+        if (RESOURCE_GROUP_QUERY_LATENCY_MAP.containsKey(resourceGroupName)) {
+            return;
+        } else {
+            GaugeMetricImpl metricMean =
+                    new GaugeMetricImpl<>(QUERY_RESOURCE_GROUP_LATENCY, Metric.MetricUnit.MILLISECONDS,
+                            "mean of resource group query latency");
+            metricMean.addLabel(new MetricLabel("type", "mean"));
+            metricMean.addLabel(new MetricLabel("name", resourceGroupName));
+            metricMean.setValue(0.0);
+            MetricRepo.addMetric(metricMean);
+
+            GaugeMetricImpl metric50Quantile =
+                    new GaugeMetricImpl<>(QUERY_RESOURCE_GROUP_LATENCY, Metric.MetricUnit.MILLISECONDS,
+                            "median of resource group query latency");
+            metric50Quantile.addLabel(new MetricLabel("type", "50_quantile"));
+            metric50Quantile.addLabel(new MetricLabel("name", resourceGroupName));
+            metric50Quantile.setValue(0.0);
+            MetricRepo.addMetric(metric50Quantile);
```

Comment:
How about using a for loop? since each metricXxxQuantile logic is completely identical except quantile value.

Reply:
I think this is just a coincidence, different metrics do not have "exactly the same" semantics. For example, one day I need to add a `tag=error_type` to `query_resource_group_err`, or filter some queries such as `select 1` for `query_resource_group`, then the metrics will be different. I think table-driven is best used for optimization of enums with "exactly the same" semantics. In this case, I think enum metrics is better.